test/robot/job/worker/action.go

```diff
@@ -95,12 +95,13 @@ func (a *Actions) apply(ctx context.Context, item Action) error {
 
 func (a *Actions) do(ctx context.Context, w *Worker, input Input) (string, error) {
 	a.mu.Lock()
-	defer a.mu.Unlock()
 	action := proto.Clone(a.nullAction).(Action)
 	action.Init(id.Unique().String(), input, w.Info)
 	if err := a.ledger.Add(ctx, action); err != nil {
 		return "", err
 	}
+	// unlock here so other do calls can be invoked after adding to ledger
```

Comment:
Ok, I was considering something like that, instead making a local function that would handle the defer. Would that be ok since this function (clone, init, add) doesn't need to be accessible anywhere else? or is the separate private function still preferred? I really like defer, but the fact that it is function oriented rather than scope oriented is still a gotcha for me.